### PR TITLE
SCI-1033: Association result annotation + Asset

### PR DESCRIPTION
**Pull request by nishill**
Please help me with code review as well as any changes that I may have missed, thanks.

Pull request also includes the project asset.

# dxapp.json
- [x] Upgrade to 14.04
- [x] Add help and extension to input spec.
- [x] In `details` field specify: `whatsNew`and `citation` (leave blank for citation.
- [x] In `details` consider specifying `upstreamAuthor` and `upstreamVersion` fields when applicable.
- [x] Use `regionalOptions` 
- [x] Set applicable timeout (Be relax) and restart policy in [`runSpec`](https://wiki.dnanexus.com/API-Specification-v1.0.0/IO-and-Run-Specifications).
- [x] `version` field, set proper [semantic versioning](https://semver.org/).
- [x] Setting `"openSource": true`. 
- [x] Run [dxapp linter](https://github.com/dnanexus/RoughER/blob/master/scripts/dxapp_lint.py) [[permalink](https://github.com/dnanexus/RoughER/blob/ecb3ea3f97870c020f93bf4723afc9f40a310e33/scripts/dxapp_lint.py)] on `dxapp.json`.
- [x] Use AssetDepend instead of execDepend
- [x] Add default instance type that works for a general use case.

# Readme.md
- [x] Improve Readme file to contain applet function, use case, input/output description, underline execution if it’s a wrapper.
- [ ] Read documentation for sanity check. (Unpublished app URL will be: https://platform.dnanexus.com/app/app-xxxx)

# Test
- [x] Basic testing using given input in Box or expected input
- [x] Make a Pull Request to `master` in https://github.com/RitchieLab/dnanexus

# Production
- [ ] Co-test with Ritchie group
- [ ] Help Ritchie group to publish (Give dx command)
- [ ] Verify the versions that appear on the platform are the intended versions
- [ ] Set up demo workflow in public project.

### CODE REVIEW

File: apps/association_result_annotation/src/association_result_annotation.sh [hunk 1]
```diff
@@ -14,251 +14,260 @@
...
+
+#################################################################################
+# Initial checks on input file ##################################################
+#################################################################################
+check_inputs() {
+
+    # check if the input file is tab seperated
+    if [ -z "$(sed -n '/\t/p;q' input_file )" ]
+    then
+        dx-jobutil-report-error "ERROR: Incorrect input file format. App only accepts tab seperated input file."
```
Comment (slzarate) [inline apps/association_result_annotation/src/association_result_annotation.sh]: Should be "separated"

File: apps/association_result_annotation/src/association_result_annotation.sh [hunk 2]
```diff
@@ -14,251 +14,260 @@
...
+        fi
+
+        # Create position file of SNPs from the association result file
+        cut -f$chr_col_num,$pos_col_num input_file |  sort -u  > input_file_snp_position.txt
+
+        # Run biofilter to get gene, upstream gene, downstream gene and GWAS
+
+        sudo mkdir /usr/share/biofilter
+        sudo chmod a+rwx /usr/share/biofilter
+
```
Comment (slzarate) [inline apps/association_result_annotation/src/association_result_annotation.sh]: Clean up whitespace

File: apps/association_result_annotation/src/association_result_annotation.sh [hunk 3]
```diff
@@ -14,251 +14,260 @@
...
-
-			# Create position file of SNPs from the association result file
-			cut -f$chr_col_num,$pos_col_num input_file |  sort -u  > input_file_snp_position.txt
-
-			# Run biofilter to get gene, upstream gene, downstream gene and GWAS
-
-			sudo mkdir /usr/share/biofilter
-			sudo chmod a+rwx /usr/share/biofilter
+    else
+        echo "Skipping Gene and GWAS Annotations!!!"
```
Comment (slzarate) [inline apps/association_result_annotation/src/association_result_annotation.sh]: Remove "!!!"

File: apps/association_result_annotation/src/association_result_annotation.sh [hunk 4]
```diff
@@ -14,251 +14,260 @@
...
+    fi
+    #################################################################################
 
 
-			dx download -r "project-Bkp5fYQ0vqZfq1XXPkYK2p1z:/Biofilter/2.4/*" -o /usr/share/biofilter/
-			dx download "project-Bkp5fYQ0vqZfq1XXPkYK2p1z:/LOKI/LOKI-20160428-noSNPs.db" -o /usr/share/biofilter/loki.db
+    #################################################################################
+    # Import input file into SQLite database ########################################
+    #################################################################################
```
Comment (slzarate) [inline apps/association_result_annotation/src/association_result_annotation.sh]: We should discuss style -- is this how we want comments to look?

**Review by slzarate — CHANGES_REQUESTED** [review]
Minor style revisions requested, otherwise LGTM

File: apps/association_result_annotation/dxapp.json [hunk 5]
```diff
@@ -1,22 +1,31 @@
...
-      "optional": false
+      "optional": false,
+      "patterns": ["*.tsv"],
+      "help": "A data file containing information from a genome wide association catalog."
     },
     {
-      "name": "sql_file",
-      "label": "SQLite file",
+      "name": "sql_db_file",
+      "label": "SQLite databse file",
```
Comment (etal) [inline apps/association_result_annotation/dxapp.json]: "database"

File: asset/tmp [hunk 6]
```diff
@@ -0,0 +1,15 @@
+SHELL=/bin/bash -e -x -o pipefail
```
Comment (etal) [inline asset/tmp]: Is this file a draft Makefile that was not meant to be committed?

Comment (nishill) [reply asset/tmp]: Yes, that's right. I will remove it now

File: apps/association_result_annotation/src/association_result_annotation.sh (hunk 4 above)
Comment (nishill) [reply apps/association_result_annotation/src/association_result_annotation.sh]: I'm open for that discussion, if I were to do block comments I would do them like this:

```
#################################################################################
#                     Odds Ratio and Confidence Interval                        #
#################################################################################
```

Comment (slzarate) [reply apps/association_result_annotation/src/association_result_annotation.sh]: @Arkarachai We should discuss this this week

**Review by slzarate — APPROVED** [review]

**Comment by slzarate** [comment]
Add Azure asset to dxapp.json

File: apps/association_result_annotation/dxapp.json [hunk 7]
```diff
@@ -1,120 +1,144 @@
...
   "name": "association_result_annotation",
   "title": "Annotate GWAS, PheWAS Associations",
-  "summary": "association_result_annotation",
+  "summary": "This app increases the depth of annotation information in an input GWAS tsv file.",
   "dxapi": "1.0.0",
-  "version": "0.0.1",
-  "categories": [],
+  "version": "1.0.0",
+  "details": {
+    "whatsNew": "*1.0.0: updated to ubuntu 14.04, added patterns/help to input/output spec, added regionalOptions/assetDepends;",
```
Comment (yihchii) [inline apps/association_result_annotation/dxapp.json:8]: "whatsNew": "\*1.0.0: updated to ubuntu 14.04, added patterns/help to input/output spec, added regionalOptions/assetDepends;",
->
"\* 1.0.0 First version of the Annotate GWAS, PheWAS Associations app."

File: apps/association_result_annotation/dxapp.json [hunk 8]
```diff
@@ -1,120 +1,144 @@
...
+    "upstreamAuthor": "Ritchie, MD",
+    "citations": []
+  },
+  "categories": ["GWAS"],
   "inputSpec": [
     {
-      "name": "input_file",
+      "name": "associations",
```
Comment (yihchii) [inline apps/association_result_annotation/dxapp.json:15]: "associations" -> "associations_tsv"
Also need to update in the app script accordingly.

File: apps/association_result_annotation/dxapp.json [hunk 9]
```diff
@@ -1,120 +1,144 @@
...
       "class": "file",
-      "optional": false
+      "optional": false,
+      "patterns": ["*.tsv"],
+      "help": "A data file containing information from a genome wide association catalog."
     },
     {
-      "name": "sql_file",
-      "label": "SQLite file",
+      "name": "sql_database",
```
Comment (yihchii) [inline apps/association_result_annotation/dxapp.json:23]: "sql_database" -> "sql_db"
Also need to update in the app script accordingly.

File: apps/association_result_annotation/dxapp.json [hunk 10]
```diff
@@ -1,120 +1,144 @@
...
-          "id": "file-BxGpVYQ0pB0vZkV6g5JVgz9F"
-      }
-      }
+      "help": "A SQLite3 databse file.",
+      "suggestions": [
+        {   
+          "project": "project-FGqF5B00ByvPXG4G87yQy23G",
+          "id": "file-FGzv6FQ0Byv9p5ZV6ZkgVv3K",
+          "path": "/GWAS_Annotations",
+          "name": "Ritchie Lab Software_dx: AWS East"
```
Comment (yihchii) [inline apps/association_result_annotation/dxapp.json:34]: May need to work with Ritchie lab to decide what "name" should be used for suggestions

File: apps/association_result_annotation/dxapp.json [hunk 11]
```diff
@@ -124,24 +148,67 @@
     {
       "name": "out_file",
```
Comment (yihchii) [inline apps/association_result_annotation/dxapp.json:149]: "out_file" -> "out_file_tsv"
Also need to update in the app script accordingly.

File: apps/association_result_annotation/dxapp.json [hunk 12]
```diff
@@ -124,24 +148,67 @@
...
-    "aws:us-east-1": {
-      "resources": "project-BYpFk1Q0pB0xzQY8ZxgJFv1V"
+    "file": "src/association_result_annotation.sh",
+    "timeoutPolicy": {
+      "main": { 
+        "hours": 12  
+      },
+      "*": {
+        "hours": 8,
+        "minutes": 30
```
Comment (yihchii) [inline apps/association_result_annotation/dxapp.json:168]: Can this be simplify to either 8 hours or 9 hours?

File: apps/association_result_annotation/dxapp.json [hunk 13]
```diff
@@ -124,24 +148,67 @@
...
+   },
+   "regionalOptions": {
+    "aws:us-east-1": {
+      "systemRequirements": {
+        "*": {
+          "instanceType": "mem1_ssd1_x4"
+        }   
+      },  
+      "assetDepends": [
+          {
```
Comment (yihchii) [inline apps/association_result_annotation/dxapp.json:180]: Please update record IDs to:
aws:us-east-1: record-FJ3zf8j08zVZFzqYG88Pbk7z
azure:westus: record-FJ4058j9Vf9ZBBVB2Q7b3VQz

File: apps/association_result_annotation/dxapp.json [hunk 14]
```diff
@@ -124,24 +148,67 @@
...
+        }
+      }
+   }   
   },
   "access": {
     "project": "CONTRIBUTE"
   },
-   "developers" : ["user-auv13", "user-atf3"],
-   "authorizedUsers": ["org-geisinger_affiliate", "org-geisinger", "org-geisinger_devel", "org-emerge_geisinger"]
+   "__developers__" : ["user-averma"],
```
Comment (yihchii) [inline apps/association_result_annotation/dxapp.json:209]: Remove \_\_developers\_\_

File: apps/association_result_annotation/dxapp.json [hunk 15]
```diff
@@ -124,24 +148,67 @@
...
+   "openSource": true,
+   "__authorizedUsers__": ["org-geisinger_affiliate", "org-geisinger", "org-geisinger_devel", "org-emerge_geisinger"],
```
Comment (yihchii) [inline apps/association_result_annotation/dxapp.json:211]: Remove \_\_authorizedUsers\_\_

**Review by yihchii — COMMENTED** [review]
Please see comments for:
1. Generalized whatsNew release log for first app release.
2. Pease follow input nomenclature.
3. Please update to use the latest version of the asset.
4. Remove unused entries

**Pull request merged**